### PR TITLE
Stories/287 288 probability distributions

### PR DESCRIPTION
**Pull request by cguardia**
WIP. Still needs some work, but making PR to ease review.

### CODE REVIEW

**Comment by vlall** [comment]
Images taken with 50x50 grid distributing 100 food each.

Random:
random

Sinusoidal
sinusoidal

Standing Wave
standing_wave

Horizontal Gradient
horizontal_gradient

Vertical Gradient
vertical_gradient

File: dlgr/griduniverse/experiment.py [hunk 1]
```diff
@@ -303,6 +307,19 @@ def __init__(self, **kwargs):
             self.color_costs = [2**i for i in range(self.num_colors)]
             random.shuffle(self.color_costs)
 
+        # get food spawning probability distribution function and args
+        self.probability_function_args = []
```
Comment (vlall) [inline dlgr/griduniverse/experiment.py]: We need some error catching for incorrect keys entered as probability distributions.

File: dlgr/griduniverse/experiment.py [hunk 2]
```diff
@@ -265,6 +268,7 @@ def __init__(self, **kwargs):
             'food_maturation_threshold', 0.0)
         self.food_planting = kwargs.get('food_planting', False)
         self.food_planting_cost = kwargs.get('food_planting_cost', 1)
+        self.food_probability_distribution = kwargs.get('food_probability_distribution', 'random')
```
Comment (vlall) [inline dlgr/griduniverse/experiment.py]: How do i add my distributions to the grid? I'm looking for the link between the `food_probability_distribution` keyword to the appropriate function

**Comment by vlall** [comment]
My goal is to add screenshots for each of the distributions to this PR, but the keywords aren't working as expected.

**Comment by vlall** [comment]
Images taken with 50x50 grid distributing 100 food each.

random:
random

sinusoidal
sinusoidal

standing_wave
standing_wave

horizontal_gradient
horizontal_gradient

vertical_gradient
vertical_gradient

File: dlgr/griduniverse/experiment.py [hunk 3]
```diff
@@ -303,6 +307,19 @@ def __init__(self, **kwargs):
             self.color_costs = [2**i for i in range(self.num_colors)]
             random.shuffle(self.color_costs)
 
+        # get food spawning probability distribution function and args
+        self.probability_function_args = []
+        parts = self.food_probability_distribution.split()
+        if len(parts) > 1:
+            self.food_probability_distribution = parts[0]
```
Comment (vlall) [inline dlgr/griduniverse/experiment.py]: We need some error catching for incorrect keys entered as probability distributions.

Comment (cguardia) [reply dlgr/griduniverse/experiment.py]: Right now, if you enter an invalid distribution name, it just is ignored, and simple random is used instead (see lines 318-320 below). No exception is raised, so the experiment always runs. How would you prefer to handle invalid names?

Comment (vlall) [reply dlgr/griduniverse/experiment.py]: I think maybe it should return a key value error, just incase somebody misspells a distribution like "sinusoidal" and it ends up running the random distribution.

Comment (cguardia) [reply dlgr/griduniverse/experiment.py]: The only problem is that the error will only be visible in the heroku logs and will result in an empty grid displayed when the experiment starts.

Comment (vlall) [reply dlgr/griduniverse/experiment.py]: Hmm, maybe some log saying that it's an incorrect key and runs with random anyway? Just so there's something in the logs.

Comment (cguardia) [reply dlgr/griduniverse/experiment.py]: Added logging.

**Comment by vlall** [comment]
I made some changes to get this story merged but I can't get any of the distributions besides the `random` one to generate food. I don't see errors in the browser console or terminal, just a blank game screen. The distributions work when plotted on a graph, I'm not sure what's breaking them.
test

**Pull request merged**